core/include/vecmem/containers/details/jagged_device_vector_iterator.hpp

```diff
@@ -0,0 +1,184 @@
+
+      /// Copy assignment operator
+      VECMEM_HOST_AND_DEVICE
+      jagged_device_vector_iterator&
+      operator=( const jagged_device_vector_iterator& rhs );
+
+      /// @name Value accessor operators
+      /// @{
+
+      /// De-reference the iterator
+      VECMEM_HOST_AND_DEVICE
+      reference operator*() const;
+      /// Use the iterator as a pointer
+      VECMEM_HOST_AND_DEVICE
+      pointer operator->() const;
+
+      /// @}
+
+      /// @name Iterator updating operators
+      /// @{
+
+      /// Decrement the underlying iterator (with '++' as a prefix)
+      VECMEM_HOST_AND_DEVICE
+      jagged_device_vector_iterator& operator++();
+      /// Decrement the underlying iterator (wuth '++' as a postfix)
+      VECMEM_HOST_AND_DEVICE
+      jagged_device_vector_iterator operator++( int );
+
+      /// Increment the underlying iterator (with '--' as a prefix)
+      VECMEM_HOST_AND_DEVICE
+      jagged_device_vector_iterator& operator--();
+      /// Increment the underlying iterator (with '--' as a postfix)
+      VECMEM_HOST_AND_DEVICE
+      jagged_device_vector_iterator operator--( int );
+
+      /// Decrement the underlying iterator by a specific value
+      VECMEM_HOST_AND_DEVICE
+      jagged_device_vector_iterator operator+( difference_type n ) const;
+      /// Decrement the underlying iterator by a specific value
+      VECMEM_HOST_AND_DEVICE
+      jagged_device_vector_iterator& operator+=( difference_type n );
+
+      /// Increment the underlying iterator by a specific value
+      VECMEM_HOST_AND_DEVICE
+      jagged_device_vector_iterator operator-( difference_type n ) const;
+      /// Increment the underlying iterator by a specific value
+      VECMEM_HOST_AND_DEVICE
+      jagged_device_vector_iterator& operator-=( difference_type n );
+
+      /// @}
+
+      /// @name Comparison operators
+      /// @{
+
+      /// Check for the equality of two iterators
+      bool operator==( const jagged_device_vector_iterator& other ) const;
+      /// Check for the inequality of two iterators
+      bool operator!=( const jagged_device_vector_iterator& other ) const;
+
+      /// @}
+
+   private:
+      /// Ensure that the internal value object is valid
+      void ensure_valid() const;
+
+      /// Pointer to the data (in an array)
+      data_pointer m_ptr;
+      /// Flag showing whether the value object is up to date with the pointer
+      mutable bool m_value_is_valid;
+      /// Helper object returned by the iterator
+      mutable value_type m_value;
+
+   }; // class jagged_device_vector_iterator
+
+} } // namespace vecmem::details
+
+namespace std {
+
+   /// Specialisation of @c std::iterator_traits
+   ///
+   /// This is necessary to make @c vecmem::reverse_iterator functional on top
+   /// of @c vecmem::details::jagged_device_vector_iterator.
+   ///
+   template< typename T >
+   struct
+   iterator_traits< vecmem::details::jagged_device_vector_iterator< T > > {
```

Comment:
I'm not completely sure about this, but I think we should also declare `iterator_traits<vecmem::details::jagged_device_vector_iterator<T>>::iterator_category` to be `random_access_iterator_tag`.

Reply:
It wouldn't hurt. It's not needed by the code, but having it there should not cause any problems either.